File: imogi_finance/imogi_finance/imogi_finance/report/vat_output_register_verified/vat_output_register_verified.js
// Copyright (c) 2024, Imogi Finance and contributors
// For license information, please see license.txt

frappe.query_reports["VAT Output Register Verified"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start(),
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_end(),
			"reqd": 0
		},
		{
			"fieldname": "customer",
			"label": __("Customer"),
			"fieldtype": "Link",
			"options": "Customer",
			"get_query": function() {
				return {
					query: "erpnext.controllers.queries.customer_query"
				};
			}
		},
		{
			"fieldname": "verification_status",
			"label": __("Verification Status"),
			"fieldtype": "Select",
			"options": ["", "Verified", "Needs Review", "Rejected"],
			"default": "Verified"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Format verification status with color indicators
		if (column.fieldname == "out_fp_status") {
			if (value === "Verified") {
				value = `<span class="indicator-pill green">${value}</span>`;
			} else if (value === "Needs Review") {
				value = `<span class="indicator-pill orange">${value}</span>`;
			} else if (value === "Rejected") {
				value = `<span class="indicator-pill red">${value}</span>`;
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Validate configuration on load
		frappe.call({
			method: "imogi_finance.imogi_finance.utils.tax_report_utils.validate_tax_register_configuration",
			args: {
				register_type: "output",
				company: frappe.query_report.get_filter_value("company")
			},
			callback: function(r) {
				if (r.message && !r.message.valid) {
					frappe.msgprint({
						title: __("Configuration Required"),
						indicator: r.message.indicator || "red",
						message: r.message.message + "<br>" + (r.message.action || "")
					});
				}
			}
		});
		
		// Add custom buttons
		report.page.add_inner_button(__("Configuration Check"), function() {
			check_configuration(report);
		}, __("Tools"));
		
		report.page.add_inner_button(__("Export to Excel"), function() {
			export_to_excel(report);
		}, __("Tools"));
	}
};

function check_configuration(report) {
	const company = frappe.query_report.get_filter_value("company");
	
	if (!company) {
		frappe.msgprint(__("Please select a company first."));
		return;
	}
	
	frappe.call({
		method: "imogi_finance.imogi_finance.utils.tax_report_utils.validate_tax_register_configuration",
		args: {
			register_type: "output",
			company: company
		},
		callback: function(r) {
			if (r.message) {
				let message = r.message.message;
				if (r.message.action) {
					message += "<br><br>" + r.message.action;
				}
				
				frappe.msgprint({
					title: __("Configuration Status"),
					indicator: r.message.indicator || "blue",
					message: message
				});
			}
		}
	});
}

function export_to_excel(report) {
	// Use native ERPNext export functionality
	const filters = frappe.query_report.get_filter_values();
	
	frappe.call({
		method: "frappe.desk.query_report.export_query",
		args: {
			report_name: report.report_name,
			file_format_type: "Excel",
			filters: filters
		},
		callback: function(r) {
			if (r.message) {
				const link = document.createElement("a");
				link.href = r.message.file_url;
				link.download = r.message.file_name;
				link.click();
			}
		}
	});
}

File: imogi_finance/imogi_finance/imogi_finance/report/vat_output_register_verified/vat_output_register_verified.py
"""
VAT Output Register Verified - ERPNext v15+ Native-First

Shows verified Sales Invoices with Output VAT details.
Only includes transactions with valid GL entries (properly posted to ledger).
"""

from __future__ import annotations

import frappe
from frappe import _
from frappe.query_builder import DocType
from frappe.query_builder.functions import Sum, Coalesce
from frappe.utils import flt, getdate
from typing import Optional, Dict, List, Any

from imogi_finance.imogi_finance.utils.tax_report_utils import (
	validate_vat_output_configuration,
	build_date_conditions,
	get_columns_with_width,
	get_tax_amount_from_gl
)


def execute(filters: Optional[Dict[str, Any]] = None) -> tuple[List[Dict], List[Dict]]:
	"""
	Execute VAT Output Register report.

	Args:
		filters: Report filters

	Returns:
		Tuple of (columns, data)
	"""
	filters = filters or {}

	# Validate configuration first
	validation = validate_vat_output_configuration(filters.get("company"))
	if not validation.get("valid"):
		frappe.msgprint(
			f"{validation.get('message')}<br>{validation.get('action', '')}",
			title=_("Configuration Error"),
			indicator=validation.get("indicator", "red"),
			raise_exception=True
		)

	ppn_output_account = validation.get("account")
	columns = get_columns()
	data = get_data(filters, ppn_output_account)

	return columns, data


def get_columns() -> List[Dict[str, Any]]:
	"""Define report columns with ERPNext v15+ standards."""
	columns = [
		{
			"label": _("Posting Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 100
		},
		{
			"label": _("Reference"),
			"fieldname": "name",
			"fieldtype": "Link",
			"options": "Sales Invoice",
			"width": 140
		},
		{
			"label": _("Customer"),
			"fieldname": "customer",
			"fieldtype": "Link",
			"options": "Customer",
			"width": 180
		},
		{
			"label": _("Buyer NPWP"),
			"fieldname": "out_fp_customer_npwp",
			"fieldtype": "Data",
			"width": 150
		},
		{
			"label": _("Tax Invoice No"),
			"fieldname": "out_fp_no",
			"fieldtype": "Data",
			"width": 180
		},
		{
			"label": _("Tax Invoice Date"),
			"fieldname": "out_fp_date",
			"fieldtype": "Date",
			"width": 110
		},
		{
			"label": _("Verification Status"),
			"fieldname": "out_fp_status",
			"fieldtype": "Data",
			"width": 130
		},
		{
			"label": _("DPP"),
			"fieldname": "out_fp_dpp",
			"fieldtype": "Currency",
			"width": 130
		},
		{
			"label": _("PPN (Invoice)"),
			"fieldname": "out_fp_ppn",
			"fieldtype": "Currency",
			"width": 130
		},
		{
			"label": _("PPN (GL Entry)"),
			"fieldname": "tax_amount_gl",
			"fieldtype": "Currency",
			"width": 140
		},
		{
			"label": _("Company"),
			"fieldname": "company",
			"fieldtype": "Link",
			"options": "Company",
			"width": 150
		},
	]

	return get_columns_with_width(columns)


def get_data(filters: Dict[str, Any], ppn_output_account: str) -> List[Dict[str, Any]]:
	"""
	Get VAT Output Register data using frappe.qb with GL Entry validation.
	Only shows invoices that have been properly posted to the ledger.
	"""
	SI = DocType("Sales Invoice")
	GL = DocType("GL Entry")

	# Build base query - JOIN with GL Entry to ensure only posted invoices are shown
	query = (
		frappe.qb.from_(SI)
		.inner_join(GL)
		.on(
			(GL.voucher_type == "Sales Invoice") &
			(GL.voucher_no == SI.name) &
			(GL.company == SI.company) &
			(GL.is_cancelled == 0)
		)
		.select(
			SI.name,
			SI.posting_date,
			SI.customer,
			SI.company,
			SI.out_buyer_tax_id,
			SI.out_fp_no,
			SI.out_fp_date,
			SI.out_fp_dpp,
			SI.out_fp_ppn,
			SI.out_fp_status
		)
		.where(SI.docstatus == 1)
		.distinct()
	)

	# Apply filters
	if filters.get("company"):
		query = query.where(SI.company == filters.get("company"))

	if filters.get("customer"):
		query = query.where(SI.customer == filters.get("customer"))

	# Verification status filter - default to "Verified" if not specified
	verification_status = filters.get("verification_status", "Verified")
	if verification_status:
		query = query.where(SI.out_fp_status == verification_status)

	# Date range conditions
	date_condition = build_date_conditions(SI, filters, "posting_date")
	if date_condition is not None:
		query = query.where(date_condition)

	# Order by posting date
	query = query.orderby(SI.posting_date).orderby(SI.name)

	# Execute query
	invoices = query.run(as_dict=True)

	# Get GL-based tax amounts for each invoice
	data = []
	for invoice in invoices:
		# Get actual tax amount from GL Entry (most reliable source)
		tax_amount_gl = get_tax_amount_from_gl(
			voucher_type="Sales Invoice",
			voucher_no=invoice.name,
			tax_account=ppn_output_account,
			company=invoice.company
		)

		# Use out_buyer_tax_id as the buyer NPWP field
		npwp = invoice.get("out_buyer_tax_id")

		data.append({
			**invoice,
			"out_fp_customer_npwp": npwp,
			"tax_amount_gl": tax_amount_gl
		})

	return data
